Bluetooth: Mesh: clean up for better integration with PSA (#84877)
zephyrproject-rtos/zephyr: doc/releases/migration-guide-4.1.rst

  set as deprecated. Default option for platforms that do not support TF-M
  is :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA`.

> * Mesh key representations are not backward compatible if images were built with TinyCrypt and
>   PSA API based crypto libraries. Mesh does not store key values anymore for PSA based crypto
>   libraries. PSA based crypto library stores keys in the internal trusted storage.
>   If a provisioned device updated its image that was built with
>   :kconfig:option:`CONFIG_BT_MESH_USES_TINYCRYPT` on image that was built with
>   :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA` or
>   :kconfig:option:`CONFIG_BT_MESH_USES_TFM_PSA` without erasing the persistent area,
>   it should be reprovisioned.

Reviewer: ```suggestion
* Mesh key representations are not backward compatible if images are built with TinyCrypt and
  crypto libraries based on the PSA API. Mesh no longer stores the key values for those crypto
  libraries. The crypto library stores the keys in the internal trusted storage.
  If a provisioned device updated its image that was built with the
  :kconfig:option:`CONFIG_BT_MESH_USES_TINYCRYPT` Kconfig option set on an image that was built with
  :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA` or
  :kconfig:option:`CONFIG_BT_MESH_USES_TFM_PSA` without erasing the persistent area,
  it should be reprovisioned.
```

>   it should be reprovisioned.

Reviewer 2: By reading this I get an impression that I can update a device running mesh with TC to mesh with PSA support and then reprovision, but this not correct. After such update mesh will not work, thus device can't be reprovisioned.

Instead we should write that the device must be unprovisioned <ins>before</ins> updating to a firmware with PSA support. And only then the device can be provisioned again. Otherwise, it will make the device unresponsive.

Reviewer 3: This seems like a pretty user-unfriendly approach. An alternative that was proposed when this was discussed in the TSC meeting, was to provide conversion code that could read values from TC-based storage and feed them to PSA, thereby creating a migration path that doesn't require unprovisioning. Is there a reason why this isn't being considered anymore?

Author: I wouldn't say it is not considered. Zephyr community so rushed with Tinycrypt deprecation that this conversion nobody implemented yet. Even more, such conversions create serious security vulnerability if the attacker has access to settings API (for example over wireless shell).
If the conversion is implemented by anybody it will be welcome. Until it the customer should be aware how to manage with migration from deprecated functionality to ongoing one.

Reviewer 3: > such conversions create serious security vulnerability if the attacker has access to settings API

Maybe I'm not understanding something here, but isn't this just as vulnerable as existing devices with TC-based storage? If you're concerned with fooling devices to do the conversion twice - couldn't you store some boolean flag in PSA storage to indicate that the conversion has already been done, thereby preventing the conversion code from running again.

Based on your skepticism with this approach, sounds like the Zephyr Mesh team isn't committing to adding such support any time in the future?

Author: > > such conversions create serious security vulnerability if the attacker has access to settings API
> 
> Maybe I'm not understanding something here, but isn't this just as vulnerable as existing devices with TC-based storage? If you're concerned with fooling devices to do the conversion twice - couldn't you store some boolean flag in PSA storage to indicate that the conversion has already been done, thereby preventing the conversion code from running again.

I'd like to avoid architecture\implementation discussion in this PR.

> 
> Based on your skepticism with this approach, sounds like the Zephyr Mesh team isn't committing to adding such support any time in the future?

We (as maintainers) do not have plans to implement it as of today. However, if any other contributors implement it, we will welcome the contribution.

  set as deprecated. Default option for platforms that do not support TF-M
  is :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA`.

* Mesh key representations are not backward compatible if images are built with TinyCrypt and
  crypto libraries based on the PSA API. Mesh no longer stores the key values for those crypto
  libraries. The crypto library stores the keys in the internal trusted storage.
  If a provisioned device is going to update its image that was built with
  the :kconfig:option:`CONFIG_BT_MESH_USES_TINYCRYPT` Kconfig option set on an image
  that was built with :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA` or
  :kconfig:option:`CONFIG_BT_MESH_USES_TFM_PSA` without erasing the persistent area,

>   it should be unprovisioned first and reprovisioned after update again.

Author: What is difference between DFU and none-DFU image update? Could you elaborate?

>   it should be unprovisioned first and reprovisioned after update again.

Reviewer 4: In case of DFU, the end-application should not forget to return the EFFECT_UNPROVE in the [callback](https://github.com/zephyrproject-rtos/zephyr/blob/a2eaddb4bcbb177a68f1feb6b43170ce31ea995f/include/zephyr/bluetooth/mesh/dfu_srv.h#L83). If they forget to do that (because they originally intend to never change their device composition across DFU updates), and if they return EFFECT_NONE, the new firmware will boot without getting unprovisioned. Then the recommendation written here will be violated. 

It will be good to avoid this situation by informing about this proactively.

>   it should be unprovisioned first and reprovisioned after update again.

Author: I guess this is not the case for DFU at all. 
```
:c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`
  This effect is chosen if the composition data in the new firmware changes, the device doesn't
  support the remote provisioning, and the new composition data takes effect after applying the
  firmware.
```
Device should be unprovisioned before update not after. Presence or absence the remote provisioning functionality does not impact anyhow.

>   it should be unprovisioned first and reprovisioned after update again.

Reviewer 4: I think we may have to update that note as well. For example:
```
:c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`
  This effect is chosen if the composition data in the new firmware changes, the device doesn't
  support the remote provisioning, and the new composition data takes effect after applying the
  firmware. This effect can also be chosen, if it is necessary to unprovision the device for 
  application specific reasons. 
```
?

>   it should be unprovisioned first and reprovisioned after update again.

Author: Shouldn't effect be chosen based on the composition data hash that is extracted from metadata?
It was the intention of the DFU algorithm. I'm pretty sure this PR recommendation is more general and shouldn't impact on any internal mesh behavior apart mandating to be unprovisioned one way or another (actually doesn't matter how).

>   it should be unprovisioned first and reprovisioned after update again.

Reviewer 2: This state in the spec doesn't mandate that CD must be changed:
![image](https://github.com/user-attachments/assets/4e2fe2f3-cfa8-41b9-b7e7-80f92e139331)

The wording used in our API aren't fully correct then.

I tend to think that yes, this is another option that could be mentioned in the migration guide as will hint customers that they can use the mesh DFU to update fw to a PSA-enabled mesh butthey should return `BT_MESH_DFU_EFFECT_UNPROV`.

>   it should be unprovisioned first and reprovisioned after update again.

Reviewer 4: Yup. Thanks for updates.

  set as deprecated. Default option for platforms that do not support TF-M
  is :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA`.

* Mesh key representations are not backward compatible if images are built with TinyCrypt and
  crypto libraries based on the PSA API. Mesh no longer stores the key values for those crypto
  libraries. The crypto library stores the keys in the internal trusted storage.
  If a provisioned device is going to update its image that was built with
  the :kconfig:option:`CONFIG_BT_MESH_USES_TINYCRYPT` Kconfig option set on an image
  that was built with :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA` or
  :kconfig:option:`CONFIG_BT_MESH_USES_TFM_PSA` without erasing the persistent area,
  it should be unprovisioned first and reprovisioned after update again.

>   If image is changed over Mesh DFU then :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV` should be used.

Reviewer: ```suggestion
  If the image is changed over Mesh DFU, use :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`.
```
